Add `post_shutdown`

ExalFabu opened the pull request:
This should resolve #3105 
I strictly followed #3078, I hope I've not missed any documentation that was required to change as well

Anyway, as I mentioned in the issue i am struggling with these test that i added:  
- `test_run_polling_post_shutdown`
- `test_run_webhook_post_shutdown`

Here's a screenshot of the error (it's the same for both tests)
![image](https://user-images.githubusercontent.com/53974096/176030688-af5a2738-6a46-4aab-8cb5-caef1dfafa16.png)

I don't know if I've messed up the test itself or I'm missing something

Bibo-Joshi commented on telegram/ext/_applicationbuilder.py:
```diff
@@ -928,6 +931,42 @@ async def post_init(application: Application) -> None:
+            .. code::
+
+                async def post_shutdown(application: Application) -> None:
+                    await db_connection.close()
```
```suggestion
                    await application.bot_data['database'].close()
```

just an idea - makes it more obvious why we might need the `application` parameter

Bibo-Joshi reviewed (changes requested):
Thanks for the PR! I left two comments below

ExalFabu commented:
I fixed and removed that duplicated line of code and the tests i wrote passed.
Though it seems another test broke: `test_run_error_in_application`, only when the error is simulated inside `application.initialize()`

And for what I understand it's really not an issue, since at this point of the execution the updater hasn't been initialized yet, and that's why `app.shutdown()` doesn't call `updater.shutdown()`

If my assumptions are correct, i may fix the test asserting `app._initialized` and `app.updater._initilized` as false, removing the need to monkeypatch the methods
Alternatively i can assert the two different cases inside a if statement
Something like `assert set(shutdowns) == ( {"application", "updater"} if method == "start" else {"application"} ) `

Bibo-Joshi commented:
The bad boy is this here:

https://github.com/python-telegram-bot/python-telegram-bot/blob/f1d03393def74bfc09f48556ea2b20e8c6098189/telegram/ext/_application.py#L374-L376

this tells `Application.shutdown` to just exit if the app wasn't properly initialized and hence `Updater.shutdown` is not called.
While we can't be 100% sure that `Updater.initialize` wasn't called by `Application.initialize`, I'm okay with that. Catching all errors that can happen during startup is a kind of a sisyphus task, so a best effort suffices IMO.

harshil21 commented on telegram/ext/_applicationbuilder.py:
```diff
@@ -928,6 +931,42 @@ async def post_init(application: Application) -> None:
+    ) -> BuilderType:
+        """
+        Sets a callback to be executed by :meth:`Application.run_polling` and
+        :meth:`Application.run_webhook` *after* executing :meth:`Application.shutdown` but
+        *before* executing :meth:`Updater.shutdown` respectively.
```
```suggestion
        :meth:`Application.run_webhook` *after* executing :meth:`Updater.shutdown`
        and :meth:`Application.shutdown`.
```

ExalFabu replied:
Oops, missed that comment

Is this up to me to commit? If yes let me know, i don't want to hold this PR just for this (sorry)

harshil21 replied:
yeah, you can just commit my suggestion (click the button above)

harshil21 reviewed (commented):
just one comment

Bibo-Joshi commented:
Thank you for the contribution :)
